src/main/zapHomeFiles/hud/tools/websockets.js

```diff
@@ -0,0 +1,126 @@
+/*
+ * WebSockets Tool
+ *
+ * Displays websocket messages in a tab at the bottom of the screen.
+ */
+
+var WebSockets = (function() {
+
+	// Constants
+	var NAME = "websockets";
+	var LABEL = I18n.t("websockets_tool");
+	var ICONS = {};
+        ICONS.CLOCK = "clock.png";	// Not actually shown anywhere, should be changed if it is going to be shown
+    var tool = {};
+
+	function initializeStorage() {
+		tool.name = NAME;
+        tool.label = LABEL;
+		tool.data = 0;
+		tool.icon = ICONS.CLOCK;
+        tool.isSelected = false;
+        tool.isHidden = true;
+		tool.panel = "";
+        tool.position = 0;
+        tool.messages = [];
+
+        utils.writeTool(tool);
+        registerForZapEvents("org.zaproxy.zap.extension.websocket.WebSocketEventPublisher");
+	}
+
+	function showOptions(tabId) {
+		var config = {};
+
+		config.tool = NAME;
+		config.toolLabel = LABEL;
+		config.options = {remove: I18n.t("common_remove")};
+
+		utils.messageFrame(tabId, "display", {action:"showButtonOptions", config:config})
+			.then(response => {
+				// Handle button choice
+				if (response.id == "remove") {
+					utils.removeToolFromPanel(tabId, NAME);
+				}
+			})
+			.catch(utils.errorHandler);
+	}
+    
+    self.addEventListener("org.zaproxy.zap.extension.websocket.WebSocketEventPublisher", event => {
+		var eventType = event.detail['event.type'];
+        
+        if (eventType == "ws.message") {
+	        let message = {};
+	        
+	        let date = new Date(Number(event.detail.timeSentInMs));
+			let dateStr = date.getHours() + ':' + date.getMinutes() + ':' + date.getSeconds() + '.' + date.getMilliseconds();
+	
+			message.timeInMs = event.detail.timeSentInMs;
+			message.time = dateStr;
+	        message.direction = event.detail.direction;
+	        message.length = event.detail.length;
+	        message.messageSummary = event.detail.messageSummary;
+	        message.opCode = event.detail.opCode + '=' + event.detail.opCodeString;
+			message.id = event.detail.messageId;
+	
+	        utils.messageAllTabs('drawer', {action: 'updateWebSockets', messages: [message]})
+	            .catch(utils.errorHandler);
+	
+			tool.messages.push(message);
```

Comment:
Interesting to note here that we don't call `utils.loadTool` before calling `utils.writeTool`. (thats how it is in the history tool as well) Technically thats how all tools should work if there wasn't that FF bug. Anyways just something to point out. I think its totally fine though.